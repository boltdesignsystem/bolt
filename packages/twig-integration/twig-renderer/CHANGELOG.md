# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [4.0.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v4.0.0-beta-4...v4.0.0) (2021-06-23)

**Note:** Version bump only for package @bolt/twig-renderer





# [3.6.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v3.5.4...v3.6.0) (2021-05-13)

**Note:** Version bump only for package @bolt/twig-renderer





# [3.4.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v3.3.1...v3.4.0) (2021-03-29)

**Note:** Version bump only for package @bolt/twig-renderer





# [3.3.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v3.2.0...v3.3.0) (2021-03-12)

**Note:** Version bump only for package @bolt/twig-renderer





## [3.1.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v3.1.1...v3.1.2) (2021-03-05)

**Note:** Version bump only for package @bolt/twig-renderer





# [3.1.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.31.2...v3.1.0) (2021-02-08)

**Note:** Version bump only for package @bolt/twig-renderer





## [2.31.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.31.0...v2.31.1) (2021-02-03)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.31.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.30.2...v2.31.0) (2021-02-02)

**Note:** Version bump only for package @bolt/twig-renderer





# [3.0.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.29.3...v3.0.0) (2021-01-08)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.23.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.22.1...v2.23.0) (2020-05-06)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.21.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.20.2...v2.21.0) (2020-04-08)


### Bug Fixes

* minor adjustment to PL patch logic ([4cc9846](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/4cc9846437f223f1c61eb3785f8c907a1b0915e1))


### Features

* apply patches to speed up Twig rendering in Pattern Lab + within the Twig Renderer ([5a21b5c](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/5a21b5c602e4735aef0e13711a17f1e5ff6f6705))





# [2.19.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.18.1...v2.19.0) (2020-03-04)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.17.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.16.3...v2.17.0) (2020-02-04)


### Features

* cherry pick install & dependency related updates from https://github.com/boltdesignsystem/bolt/pull/1711 ([dfa3dd0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/dfa3dd08d19a109736f64faef851e5516e3941e0))





# [2.16.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.15.2...v2.16.0) (2020-01-24)

**Note:** Version bump only for package @bolt/twig-renderer





## [2.15.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.15.1...v2.15.2) (2020-01-21)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.15.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.14.3...v2.15.0) (2020-01-17)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.14.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.13.3...v2.14.0) (2020-01-06)


### Bug Fixes

* update testing-helpers package path ([9b431b6](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/9b431b6))





# [2.13.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.12.1...v2.13.0) (2019-12-13)


### Features

* upgrade PHP dependencies + update to use v0.13 of the Twig renderer via https://github.com/pattern-lab/patternlab-node/pull/1107 ([b12a6ff](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/b12a6ff))





# [2.12.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.11.4...v2.12.0) (2019-11-26)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.11.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.10.0...v2.11.0) (2019-11-14)

**Note:** Version bump only for package @bolt/twig-renderer





## [2.9.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.9.0...v2.9.1) (2019-10-22)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.9.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.8.3...v2.9.0) (2019-10-22)


### Bug Fixes

* address install / symlink issues by updating composer.json configs to allow local beta versions of Bolt packages to get installed ([77addbd](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/77addbd))
* update PHP-related packages to be in sync with the latest version of Bolt + symlinked properly ([32be05b](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/32be05b))





# [2.8.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.8.0-beta.6...v2.8.0) (2019-10-11)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.8.0-beta.6](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.8.0-beta.5...v2.8.0-beta.6) (2019-10-04)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.8.0-beta.4](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.8.0-beta.3...v2.8.0-beta.4) (2019-09-26)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.8.0-beta.3](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.7.1...v2.8.0-beta.3) (2019-09-24)



# [2.8.0-beta.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.7.0...v2.8.0-beta.2) (2019-09-18)


### Bug Fixes

* **twig-renderer:** upgrade @basalt/twig-renderer 0.12.0 => 0.12.1 ([d57414f](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/d57414f))





# [2.8.0-beta.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.7.0...v2.8.0-beta.2) (2019-09-18)


### Bug Fixes

* **twig-renderer:** upgrade @basalt/twig-renderer 0.12.0 => 0.12.1 ([d57414f](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/d57414f))





# [2.8.0-beta.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.7.0...v2.8.0-beta.1) (2019-09-14)


### Bug Fixes

* **twig-renderer:** upgrade @basalt/twig-renderer 0.12.0 => 0.12.1 ([d57414f](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/d57414f))





# [2.7.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.6.0...v2.7.0) (2019-09-13)


### Bug Fixes

* upgrade Twig renderer to the latest version to fix occasional memory errors getting thrown ([d0bdb39](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/d0bdb39))





# [2.6.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.6.0-beta.2...v2.6.0) (2019-08-30)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.6.0-beta.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.6.0-beta.1...v2.6.0-beta.2) (2019-08-27)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.6.0-beta.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.5.6...v2.6.0-beta.1) (2019-08-09)

**Note:** Version bump only for package @bolt/twig-renderer





## [2.5.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.5.1...v2.5.2) (2019-06-25)

**Note:** Version bump only for package @bolt/twig-renderer





## [2.5.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.5.0...v2.5.1) (2019-06-21)

**Note:** Version bump only for package @bolt/twig-renderer





## [2.4.4](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.4.3...v2.4.4) (2019-06-05)


### Bug Fixes

* split up the core-php BoltExtra Twig extensions into the ones needed for Drupal compatibility vs any extra Twig extensions used internally for building the docs site. Update Twig renderer configs to use the full set of Twig extensions for the docs site but only BoltCore and BoltCoreCompat when running component tests ([0287e7a](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/0287e7a))





## [2.4.3](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.4.2...v2.4.3) (2019-05-31)

**Note:** Version bump only for package @bolt/twig-renderer





# [2.4.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.3.2...v2.4.0) (2019-05-14)


### Bug Fixes

* version bump composer.lock files to fix Twig version resolved ([6f6485c](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/6f6485c))
* version bump core-php getting used + update PHP dependencies so the casehelper package is installed as expected ([9808cb4](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/9808cb4))



# [2.4.0-beta.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.2.2...v2.4.0-beta.0) (2019-04-16)


### Bug Fixes

* re-add local PHP dependencies to fix broken CI tests ([595c19d](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/595c19d))
* remove deps not needed ([4d1098c](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/4d1098c))
* retest docker build ([f4bfa64](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/f4bfa64))
* revert keep alive config ([42c255f](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/42c255f))
* revert twig-renderer vendor path updates ([28cacb8](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/28cacb8))
* update npm scripts to not use postinstall automatically; add postinstall to root of the repo instead ([5dc3496](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/5dc3496))


### Features

* **website, twig renderer, component explorer:** update the Twig Renderer to support keepAlive ([470f7af](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/470f7af))
* wire up button component to include VRT, web component tests, update Twig renderer calls to compile even faster ([4b34184](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/4b34184))
* wire up server side rendering process to core-php + add logic to the bolt text and button components to handle the differences in rendering ([752c0df](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/752c0df))





# [2.4.0-beta.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.3.0...v2.4.0-beta.0) (2019-04-16)


### Bug Fixes

* re-add local PHP dependencies to fix broken CI tests ([595c19d](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/595c19d))
* remove deps not needed ([4d1098c](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/4d1098c))
* retest docker build ([f4bfa64](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/f4bfa64))
* revert keep alive config ([42c255f](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/42c255f))
* revert twig-renderer vendor path updates ([28cacb8](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/28cacb8))
* update npm scripts to not use postinstall automatically; add postinstall to root of the repo instead ([5dc3496](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/5dc3496))


### Features

* **website, twig renderer, component explorer:** update the Twig Renderer to support keepAlive ([470f7af](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/470f7af))
* wire up button component to include VRT, web component tests, update Twig renderer calls to compile even faster ([4b34184](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/4b34184))
* wire up server side rendering process to core-php + add logic to the bolt text and button components to handle the differences in rendering ([752c0df](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/752c0df))





# [2.3.0-rc.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.2.2...v2.3.0-rc.0) (2019-01-08)


### Bug Fixes

* fix local Bolt package versions ([f1a4647](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/f1a4647))



## [2.2.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.2.1...v2.2.2) (2019-01-07)


### Bug Fixes

* re-enable schema validation for pattern lab but keep disabled for the docs site till 500 server error is troubleshooted ([ada76ef](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/ada76ef))



## [2.2.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.2.0...v2.2.1) (2018-12-17)



# [2.2.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.2.0-rc.1...v2.2.0) (2018-11-18)



# [2.2.0-rc.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.1.6...v2.2.0-rc.1) (2018-11-07)



## [2.1.6](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.1.5...v2.1.6) (2018-10-23)


### Bug Fixes

* fix long-standing issue with static docs site not recompiling when files have changed (while in watch mode); improve static site build reliability ([5928ee7](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/5928ee7))



## [2.1.4](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.1.3...v2.1.4) (2018-10-18)



## [2.1.3](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.1.2...v2.1.3) (2018-10-18)



## [2.1.2](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.1.1...v2.1.2) (2018-10-16)



## [2.1.1](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.1.0...v2.1.1) (2018-10-10)



# [2.1.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.1.0-beta.0...v2.1.0) (2018-10-10)



# [2.1.0-beta.0](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/compare/v2.0.0...v2.1.0-beta.0) (2018-10-05)


### Bug Fixes

* remove snyk check for now -- causing publishing issues ([bcf7e6b](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/bcf7e6b))


### Features

* **twig-renderer:** creates @bolt/twig-renderer for rendering individual Twig components ([277b7df](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/277b7df))


### Performance Improvements

* **twig-renderer:** improving speed of twig-renderer ([0e3b44f](https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer/commit/0e3b44f))
