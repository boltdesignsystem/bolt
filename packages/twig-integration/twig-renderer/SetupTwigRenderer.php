<?php

require_once 'vendor/autoload.php';

function addBoltCoreExtensions(\Twig\Environment &$env, $config) {
  $env->addExtension(new \Bolt\TwigExtensions\BoltCore());
  $env->addExtension(new \Bolt\TwigExtensions\BoltCoreCompat());
  // $env->addExtension(new \Twig_Extension_Debug());

  $env->addFunction(new \Twig_SimpleFunction('customTwigFunctionThatSaysWorld', function () {
    return 'Custom World';
  }));

  // temporarily disable Twig schema validation for the static site generator till the unknown 500 error being thrown is troubleshooted
  $env->addGlobal('enable_json_schema_validation', false);
}

function addBoltExtraExtensions(\Twig\Environment &$env, $config) {
  $env->addExtension(new \Bolt\TwigExtensions\BoltExtras());
}
