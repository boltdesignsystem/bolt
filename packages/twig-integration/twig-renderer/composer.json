{
  "name": "bolt-design-system/twig-renderer",
  "type": "library",
  "description": "Twig rendering service used by the Bolt Design System.",
  "version": "2.20.0",
  "license": "MIT",
  "repositories": [
    {
      "type": "path",
	    "url": "../twig-extensions-shared"
	  },
    {
      "type": "path",
      "url": "../twig-extensions-compat"
    },
    {
      "type": "vcs",
      "url": "https://github.com/boltdesignsystem/twig-tools",
      "no-api": true
    }
  ],
  "require": {
    "bolt-design-system/core-php": "*",
    "bolt-design-system/drupal-twig-extensions": "*",
    "twig/twig": "^1.42.5 || ^2.0",
    "webmozart/path-util": "*",
    "symfony/finder": "*"
  },
  "scripts": {
    "setup": "@composer install --no-interaction --prefer-dist --no-progress"
  },
  "minimum-stability": "dev"
}
