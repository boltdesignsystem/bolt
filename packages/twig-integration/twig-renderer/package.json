{
  "name": "@bolt/twig-renderer",
  "version": "4.0.0",
  "description": "Renders Twig templates using an on-the-fly PHP rendering service.",
  "keywords": [
    "design system",
    "bolt design system",
    "twig",
    "php"
  ],
  "homepage": "https://boltdesignsystem.com",
  "bugs": "https://github.com/bolt-design-system/bolt/issues",
  "repository": "https://github.com/bolt-design-system/bolt/tree/master/packages/twig-renderer",
  "license": "MIT",
  "main": "twig-renderer.js",
  "scripts": {
    "setup": "composer run setup"
  },
  "dependencies": {
    "@basalt/twig-renderer": "boltdesignsystem/basalt-twig-renderer#master",
    "@bolt/build-utils": "^4.0.0",
    "@bolt/drupal-twig-extensions": "^4.0.0",
    "sleep-promise": "^8.0.1"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependenciesMeta": {
    "@basalt/twig-renderer": {
      "unplugged": true
    }
  },
  "gitHead": "3d198feea5f73c63913a956a2a077e229ad8d462",
  "maintainers": [
    {
      "name": "Salem Ghoweri",
      "email": "me@salemghoweri.com",
      "web": "https://github.com/sghoweri"
    }
  ],
  "schema": "utils/config.schema.yml"
}
