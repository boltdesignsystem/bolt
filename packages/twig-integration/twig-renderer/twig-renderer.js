const TwigRenderer = require('@basalt/twig-renderer');
const sleep = require('sleep-promise');
const { getTwigNamespaceConfig } = require('@bolt/build-utils/manifest');
const { getConfig } = require('@bolt/build-utils/config-store');
const path = require('path');

let twigNamespaces;
let twigRenderer;

const STATES = {
  NOT_STARTED: 'NOT_STARTED',
  STARTING: 'STARTING',
  READY: 'READY',
};
let state = STATES.NOT_STARTED;

/**
 * Initialize Twig Renderer instance
 * @param {Boolean} keepAlive - Optionally tell the Twig renderer service to keep alive to help speed up requests
 */
async function init(keepAlive = false) {
  state = STATES.STARTING;
  const config = await getConfig();
  const relativeFrom = path.dirname(config.configFileUsed);
  // console.log({ config });
  twigNamespaces = await getTwigNamespaceConfig(
    relativeFrom,
    config.extraTwigNamespaces,
  );
  twigRenderer = new TwigRenderer({
    relativeFrom,
    src: {
      roots: [relativeFrom],
      namespaces: TwigRenderer.convertLegacyNamespacesConfig(twigNamespaces),
    },
    debug: false,
    alterTwigEnv: config.alterTwigEnv,
    hasExtraInfoInResponses: false, // Will add `info` onto results with a lot of info about Twig Env
    maxConcurrency: 50,
    keepAlive, // only set this to be true when doing in-browser requests to avoid issues with this process not exiting when complete
  });
  state = STATES.READY;
}

/**
 * Stops any currently running Twig Renderer instances
 */
async function stop() {
  twigRenderer.stop();
}

async function prep(keepAlive) {
  switch (state) {
    case STATES.READY:
      return;
    case STATES.NOT_STARTED:
      return await init(keepAlive);
    case STATES.STARTING:
      while (state === STATES.STARTING) {
        await sleep(20); // eslint-disable-line no-await-in-loop
      }
  }
}

/**
 * Render Twig Template
 * @param {string} template - Template name (i.e. `@bolt-components-button/button.twig`)
 * @param {Object} data - Optional data to pass to template
 * @param {Boolean} keepAlive - Optionally tell the Twig renderer service to keep alive to help speed up requests
 * @return {Promise<{ ok: boolean, html: string, message: string }>} - Results of render
 */
async function render(template, data = {}, keepAlive = false) {
  await prep(keepAlive);
  const results = await twigRenderer.render(template, data);
  return results;
}

/**
 * Render Twig String
 * @param {string} templateString - String that is a Twig template (i.e. `<p>{{ text }}</p>`)
 * @param {Object} data - Optional data to pass to template
 * @param {Boolean} keepAlive - Optionally tell the Twig renderer service to keep alive to help speed up requests
 * @return {Promise<{ ok: boolean, html: string, message: string }>} - Results of render
 */
async function renderString(templateString, data = {}, keepAlive = false) {
  await prep(keepAlive);
  const results = await twigRenderer.renderString(templateString, data);
  // console.log({ results });
  return results;
}

// Both `render` and `renderString` will return something like this on success:
//   {
//     ok: true,
//     html: '<p>Hello World</p>',
//   }
// And something like this on failure:
//   {
//     ok: false,
//     message: 'The confabulator was not configured correctly',
//   }

module.exports = {
  render,
  renderString,
  stop,
};
