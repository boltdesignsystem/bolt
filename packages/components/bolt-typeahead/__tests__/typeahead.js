import { renderString, stop as stopTwigRenderer } from '@bolt/twig-renderer';

async function renderTwigString(template, data) {
  return await renderString(template, data, true);
}

const timeout = 120000;

describe('<bolt-typeahead> Component', () => {
  let page;

  beforeEach(async () => {
    await page.evaluate(() => {
      document.body.innerHTML = '';
    });
  }, timeout);

  beforeAll(async () => {
    page = await global.__BROWSER__.newPage();
    await page.goto('http://127.0.0.1:4444/', {
      timeout: 0,
    });
  }, timeout);

  afterAll(async () => {
    await stopTwigRenderer();
    await page.close();
  }, timeout);

  test('Typeahead Twig Template Renders', async () => {
    const results = await renderTwigString(`
      {% include "@bolt-components-form/form.twig" with {
        children: include("@bolt-components-typeahead/typeahead.twig", {
          attributes: {
            class: [
              "js-typeahead-hook"
            ]
          },
          max_results: 5,
        }),
        attributes: {
          action: "https://www.pega.com/search",
          target: "_blank",
          method: "GET"
        }
      } %}
    `);
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });
});
