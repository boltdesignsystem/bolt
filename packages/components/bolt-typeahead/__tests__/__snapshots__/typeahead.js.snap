// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<bolt-typeahead> Component Typeahead Twig Template Renders 1`] = `
<form action="https://www.pega.com/search"
      target="_blank"
      method="GET"
>
  <bolt-typeahead class="js-typeahead-hook"
                  max-results="5"
  >
    <div class="c-bolt-typeahead">
      <label for="12345"
             class="c-bolt-typeahead__label"
      >
        <span class="u-bolt-visuallyhidden">
        </span>
      </label>
      <input id="12345"
             type="text"
             autocomplete="off"
             aria-autocomplete="list"
             class="c-bolt-typeahead__input"
             placeholder="Enter your search query"
             value
             required
             name
      >
      <button type="reset"
              onclick="this.previousElementSibling.focus()"
              class="c-bolt-typeahead__button c-bolt-typeahead__button--clear e-bolt-button e-bolt-button--transparent e-bolt-button--icon-only"
              aria-label="Clear search results"
      >
        <span class="e-bolt-button__icon-center"
              aria-hidden="true"
        >
          <bolt-icon class="c-bolt-typeahead__icon"
                     name="close-solid"
                     aria-hidden="true"
          >
            <span class="c-bolt-icon c-bolt-icon--close-solid">
              <svg class="c-bolt-icon__icon"
                   xmlns="http://www.w3.org/2000/svg"
                   data-name="Layer 1"
                   viewbox="0 0 22 22"
              >
                <path fill="var(--bolt-theme-icon, currentColor)"
                      d="M11 0a11 11 0 1011 11A11 11 0 0011 0zm3.5 13.1a1 1 0 010 1.4 1 1 0 01-1.4 0l-2.1-2-2.1 2a1 1 0 01-1.4 0 1 1 0 010-1.4l2-2.1-2-2.1a1 1 0 011.4-1.4l2.1 2 2.1-2a1 1 0 011.4 1.4l-2 2.1z"
                >
                </path>
              </svg>
            </span>
          </bolt-icon>
        </span>
      </button>
      <div class="c-bolt-typeahead__button c-bolt-typeahead__button--submit">
        <button type="submit"
                onclick="this.previousElementSibling.focus()"
                class="c-bolt-typeahead__button c-bolt-typeahead__button--clear e-bolt-button e-bolt-button--transparent e-bolt-button--icon-only"
                aria-label="Submit search query"
        >
          <span class="e-bolt-button__icon-center"
                aria-hidden="true"
          >
            <bolt-icon class="c-bolt-typeahead__icon"
                       name="search"
                       aria-hidden="true"
            >
              <span class="c-bolt-icon c-bolt-icon--search">
                <svg class="c-bolt-icon__icon"
                     xmlns="http://www.w3.org/2000/svg"
                     viewbox="0 0 24 24"
                >
                  <g fill="none"
                     fill-rule="evenodd"
                  >
                    <path>
                    </path>
                    <path fill="var(--bolt-theme-icon, currentColor)"
                          d="M9.5 15a5.5 5.5 0 110-11 5.5 5.5 0 010 11m12.2 5.3L15.5 14c1-1.3 1.5-2.9 1.5-4.6a7.5 7.5 0 10-3 6l6.3 6.2a1 1 0 001.4-1.4"
                    >
                    </path>
                  </g>
                </svg>
              </span>
            </bolt-icon>
          </span>
        </button>
      </div>
    </div>
  </bolt-typeahead>
</form>
`;
